File: middleware/tool_filter_middleware.py
from __future__ import annotations

import logging
from collections import defaultdict
from typing import Any, Callable, DefaultDict, Dict, Optional, Protocol, Set, cast

from fastmcp.server.dependencies import get_http_headers
from fastmcp.server.middleware import Middleware, MiddlewareContext

logger = logging.getLogger(__name__)

# attribute name to store registry on the mcp instance
_REGISTRY_ATTR = "_toolset_registry"

# ============================================================================
# Type Definitions
# ============================================================================

ToolsetRegistry = DefaultDict[str, Set[str]]  # toolset -> {tool names}

# ============================================================================
# Registry Management
# ============================================================================


def get_registry(mcp: MCPInstance) -> ToolsetRegistry:
    """Return the per-server toolset registry attached to the FastMCP instance."""
    try:
        reg = getattr(mcp, _REGISTRY_ATTR, None)
        if reg is None:
            reg = defaultdict(set)
            setattr(mcp, _REGISTRY_ATTR, reg)
        return cast(ToolsetRegistry, reg)
    except (AttributeError, TypeError) as e:
        logger.error(f"Failed to access/create registry on MCP instance: {e}")
        # Return empty registry as fallback
        return defaultdict(set)


class MCPInstance(Protocol):
    # ============================================================================
    # Decorator
    # ============================================================================
    def tool(
        self, **tool_kwargs: Any
    ) -> Callable[[Callable[..., Any]], Callable[..., Any]]: ...


def tool(mcp: MCPInstance, *, toolset: str, **tool_kwargs: Any):
    """
    Decorator that registers a tool with a toolset for HTTP header-based filtering.

    This wraps mcp.tool(...) and automatically registers the tool in the specified
    toolset, enabling clients to request specific groups of tools via the
    x-mcp-tools HTTP header.

    Args:
        mcp: The FastMCP server instance
        toolset: The toolset name this tool belongs to (e.g., "monitoring", "policy")
                Clients can request all tools in a toolset by including the toolset
                name in the x-mcp-tools header
        **tool_kwargs: All standard mcp.tool() parameters (name, description, etc.)

    Usage:
        @tool(mcp, toolset="monitoring", name="meraki_monitor_clients",
            description="Monitor clients on a network")
        def meraki_monitor_clients(...):
            ...

    Header filtering:
        - x-mcp-tools: monitoring → Returns all tools in "monitoring" toolset
        - x-mcp-tools: monitoring,policy → Returns tools from both toolsets
        - x-mcp-tools: meraki_monitor_clients → Returns specific tool only
    """

    def decorator(fn: Callable):
        wrapped = mcp.tool(**tool_kwargs)(fn)
        tool_name = (
            getattr(wrapped, "name", None) or tool_kwargs.get("name") or fn.__name__
        )
        get_registry(mcp)[toolset].add(tool_name)
        return wrapped

    return decorator


class ToolFilterMiddleware(Middleware):
    """
    GitHub-style tool filtering middleware.
    Uses per-mcp toolset registry (no hardcoding).
    """

    def __init__(self, mcp: MCPInstance, header_name: str = "x-mcp-tools") -> None:
        self.header_name = header_name
        self._mcp = mcp  # capture the instance this middleware belongs to

    def _expand_toolsets(self, names: Set[str]) -> Set[str]:
        registry = get_registry(self._mcp)
        expanded: Set[str] = set()

        for name in names:
            # if value matches a toolset, expand it; else treat it as a tool name
            if name in registry:
                logger.debug(f"Expanding toolset '{name}' -> {registry[name]}")
                expanded.update(registry[name])
            else:
                logger.debug(f"Treating '{name}' as individual tool")
                expanded.add(name)

        return expanded

    def _get_header_value(self) -> Optional[str]:
        """Return raw header value, or None if header is missing or error occurs."""
        try:
            headers = get_http_headers(include_all=True)
        except Exception as e:
            logger.error(
                "ToolFilterMiddleware: failed to read HTTP headers: %s",
                e,
                exc_info=True,
            )
            # Consistent with _get_header_candidates: fail-safe by denying
            # Or change both to allow-all for better UX
            return None  # Current: allow-all

        for k, v in headers.items():
            if k.casefold() == self.header_name.casefold():
                return v
        return None

    def _get_header_candidates(self) -> Optional[Set[str]]:
        """
        Header semantics:
          - Missing => allow ALL tools (None)
          - Present but empty => allow NONE (empty set)
          - Otherwise => allow listed tools/toolsets
        """
        try:
            raw = self._get_header_value()

            if raw is None:
                return None  # allow all

            raw = raw.strip()
            if raw == "":
                return set()  # allow none

            items = {x.strip() for x in raw.split(",") if x.strip()}
            if not items:
                return set()

            return self._expand_toolsets(items)

        except Exception as e:
            logger.error(
                "ToolFilterMiddleware._get_header_candidates failed: %s",
                e,
                exc_info=True,
            )
            return set()  # safe deny-all

    async def on_list_tools(self, context: MiddlewareContext, call_next):
        tools = await call_next(context)
        header_candidates = self._get_header_candidates()

        if header_candidates is None:
            return tools

        if not header_candidates:
            return []

        allowed_names = {t.name for t in tools} & header_candidates
        if not allowed_names:
            return []

        return [t for t in tools if t.name in allowed_names]

File: middleware/typescript/tool-filter-middleware.ts
/**
 * Generic Tool Filter Middleware for MCP Servers
 * 
 * Filters tools based on X-MCP-Tools header, supporting:
 * - Toolsets: repository, cicd, integration, users, monitoring, etc.
 * - Individual tools: server_specific_tool_name, another_tool_name
 * - Mixed: repository,specific_tool_name,cicd
 * 
 * Design Pattern:
 * - Header missing => allow ALL tools
 * - Header empty => allow NO tools  
 * - Header with values => allow only specified toolsets + individual tools
 * 
 * This middleware is reusable across all TypeScript MCP servers.
 */

export interface Tool {
  name: string;
  description: string;
  [key: string]: any;
}

export interface ToolFilterOptions {
  /**
   * The header name to look for (default: 'x-mcp-tools')
   */
  headerName?: string;
  
  /**
   * Server name for logging (e.g., 'GitLab', 'Meraki', 'AWS')
   */
  serverName?: string;
  
  /**
   * Function to get toolsets mapping (toolset name -> Set of tool names)
   */
  getToolsets: () => Record<string, Set<string>>;
  
  /**
   * Function to validate if a name is a valid individual tool
   */
  isValidTool: (toolName: string) => boolean;
}

export class ToolFilterMiddleware {
  private readonly headerName: string;
  private readonly serverName: string;
  private readonly getToolsets: () => Record<string, Set<string>>;
  private readonly isValidTool: (toolName: string) => boolean;

  constructor(options: ToolFilterOptions) {
    this.headerName = options.headerName || 'x-mcp-tools';
    this.serverName = options.serverName || 'MCP';
    this.getToolsets = options.getToolsets;
    this.isValidTool = options.isValidTool;
  }

  /**
   * Expand toolset names into individual tool names.
   * 
   */
  private expandToolsets(names: Set<string>): Set<string> {
    const expanded = new Set<string>();

    try {
      const toolsets = this.getToolsets();

      // Validate toolsets is a valid object
      if (!toolsets || typeof toolsets !== 'object') {
        console.error(`[${this.serverName}] Invalid toolsets data returned`);
        return expanded; // Return empty set as fallback
      }

      // Process each name with individual error handling
      for (const name of names) {
        try {
          const toolset = toolsets[name];

          if (toolset && toolset.size > 0) {
            // It's a toolset with tools - expand to multiple tools
            console.log(`[${this.serverName}] Expanding toolset '${name}' -> ${[...toolset].join(', ')}`);
            toolset.forEach(tool => expanded.add(tool));
          } else {
            // Check if it's a valid individual tool name
            if (this.isValidTool(name)) {
              // It's a valid individual tool - add it
              console.log(`[${this.serverName}] Adding individual tool '${name}'`);
              expanded.add(name);
            } else {
              // Unknown name - log warning but don't fail
              console.warn(`[${this.serverName}] Unknown toolset or tool name: '${name}'`);
            }
          }
        } catch (itemError) {
          // Log error for this specific item but continue processing others
          console.error(`[${this.serverName}] Error processing item '${name}':`, itemError);
          // Continue processing other items
        }
      }
    } catch (error) {
      console.error(`[${this.serverName}] ToolFilterMiddleware.expandToolsets failed:`, error);
    }

    return expanded;
  }

  /**
   * Extract and parse the X-MCP-Tools header value.
   * Returns null if header missing (allow all), Set if present
   */
  private getHeaderCandidates(headers: Record<string, any>): Set<string> | null {
    try {
      // Normalize headers to lowercase for case-insensitive lookup
      const normalizedHeaders: Record<string, any> = {};
      Object.keys(headers).forEach(key => {
        normalizedHeaders[key.toLowerCase()] = headers[key];
      });

      const rawValue = normalizedHeaders[this.headerName.toLowerCase()];

      // Header missing or undefined => allow ALL tools
      if (rawValue === undefined) {
        console.log(`[${this.serverName}] Header '${this.headerName}' missing => allow ALL tools`);
        return null;
      }

      const trimmedValue = String(rawValue).trim();

      // Header present but empty => allow NO tools
      if (trimmedValue === '') {
        console.log(`[${this.serverName}] Header '${this.headerName}' empty => allow NO tools`);
        return new Set();
      }

      // Parse comma-separated values
      const items = new Set(
        trimmedValue.split(',')
          .map(item => item.trim())
          .filter(item => item.length > 0)
      );

      if (items.size === 0) {
        return new Set();
      }

      console.log(`[${this.serverName}] Header '${this.headerName}' items: ${[...items].join(', ')}`);
      return this.expandToolsets(items);

    } catch (error) {
      console.error(`[${this.serverName}] ToolFilterMiddleware.getHeaderCandidates failed:`, error);
      // Safe fallback: deny all tools
      return new Set();
    }
  }

  /**
   * Filter tools based on X-MCP-Tools header.
   */
  public filterTools(tools: Tool[], headers: Record<string, any> = {}): Tool[] {
    try {
      const headerCandidates = this.getHeaderCandidates(headers);

      // Header missing => return all tools
      if (headerCandidates === null) {
        console.log(`[${this.serverName}] No tool filtering - returning all ${tools.length} tools`);
        return tools;
      }

      // Header empty => return no tools
      if (headerCandidates.size === 0) {
        console.log(`[${this.serverName}] Empty tool filter - returning 0 tools`);
        return [];
      }

      // Filter tools by allowed names
      const allowedNames = new Set([...tools.map(t => t.name)].filter(name => headerCandidates.has(name)));
      
      if (allowedNames.size === 0) {
        console.log(`[${this.serverName}] No matching tools found - returning 0 tools`);
        return [];
      }

      const filteredTools = tools.filter(tool => allowedNames.has(tool.name));
      console.log(`[${this.serverName}] Filtered to ${filteredTools.length} tools: ${[...allowedNames].join(', ')}`);
      
      return filteredTools;

    } catch (error) {
      console.error(`[${this.serverName}] ToolFilterMiddleware.filterTools failed:`, error);
      throw error;
    }
  }

  /**
   * Check if a specific tool is allowed by current headers
   */
  public isToolAllowed(toolName: string, headers: Record<string, any> = {}): boolean {
    const headerCandidates = this.getHeaderCandidates(headers);
    
    // Header missing => allow all
    if (headerCandidates === null) {
      return true;
    }
    
    // Header empty => allow none
    if (headerCandidates.size === 0) {
      return false;
    }
    
    return headerCandidates.has(toolName);
  }

  /**
   * Get statistics about filtering
   */
  public getFilterStats(headers: Record<string, any> = {}): {
    headerPresent: boolean;
    allowAll: boolean;
    allowNone: boolean;
    allowedTools: string[];
    serverName: string;
  } {
    const headerCandidates = this.getHeaderCandidates(headers);
    
    return {
      headerPresent: headerCandidates !== null,
      allowAll: headerCandidates === null,
      allowNone: headerCandidates !== null && headerCandidates.size === 0,
      allowedTools: headerCandidates ? [...headerCandidates] : [],
      serverName: this.serverName,
    };
  }
}

/**
 * Factory function to create a configured tool filter middleware
 */
export function createToolFilterMiddleware(options: ToolFilterOptions): ToolFilterMiddleware {
  return new ToolFilterMiddleware(options);
}

/**
 * Type definitions for better TypeScript integration
 */
export type ToolFilterStatsResult = ReturnType<ToolFilterMiddleware['getFilterStats']>;

/**
 * Helper function for common FastMCP integration pattern
 */
export function createFilteredTool<T extends Tool>(
  tool: T,
  middleware: ToolFilterMiddleware
): T & { canAccess: (auth: any) => boolean } {
  return {
    ...tool,
    canAccess: (auth: any) => {
      const headers = auth?.headers || {};
      const allowed = middleware.isToolAllowed(tool.name, headers);
      return allowed;
    }
  };
}
